components/mysql/sources/new-row-custom-query/new-row-custom-query.js

```diff
@@ -0,0 +1,70 @@
+const common = require("../common.js");
+
+module.exports = {
+  ...common,
+  key: "mysql-new-row-custom-query",
+  name: "New Row (Custom Query)",
+  description: "Emits an event when new rows are returned from a custom query",
+  version: "0.0.1",
+  dedupe: "unique",
+  props: {
+    ...common.props,
+    db: "$.service.db",
+    table: { propDefinition: [common.props.mysql, "table"] },
+    column: {
+      propDefinition: [
+        common.props.mysql,
+        "column",
+        (c) => ({ table: c.table }),
+      ],
+      label: "Order By",
+      description:
+        "A datetime column, such as 'date_updated' or 'last_modified' that is set to the current datetime when a row is updated.",
+    },
+    query: { propDefinition: [common.props.mysql, "query"] },
+  },
+  hooks: {
+    async deploy() {
+      const connection = await this.mysql.getConnection();
+      let query = `SELECT * FROM ${this.table} ORDER BY ${this.column} DESC LIMIT 10`;
+      query = this.createJoinQuery(query, this.column);
+      const rows = await this.mysql.executeQuery(connection, query);
+      this._setLastResult(rows, this.column);
+      this.iterateAndEmitEvents(rows);
+
+      await this.mysql.closeConnection(connection);
+    },
+  },
+  methods: {
+    ...common.methods,
+    /**
+     * In this method, we create an INNER JOIN to combine the given query with a query to limit results to either
+     * the 10 max results or only results updated since the lastResult.
+     * @param {object} connection - The database connection.
+     * @param {string} column - The table column used to ORDER BY
+     */
+    createJoinQuery(query, column) {
+      return `
+        SELECT * FROM (${query}) t1
+        INNER JOIN
+        (${this.query}) t2
+        ON t1.${column} = t2.${column};
+      `;
+    },
```

Comment:
I set it up this way to limit the number of incoming results (to 10) on the first run, but it might be better to leave the user's query alone and rely on the user not to enter anything that will bring in too many records. We should still limit it by the lastResult column though, correct? That's a restriction imposed on all the mysql components involving new/updated rows. Let me know what you think.

Reply:
I was thinking of a case when a user would like to execute some arbitrary query, with arbitrary payload for which we cannot apply any sorting and/or filtering (nor should we).

For instance, the user might already be applying the filtering in the SQL query itself, and the entire result should be processed here:

```sql
SELECT
  `email`,
  `first_name`,
  `last_name`
FROM `users`
WHERE
  DATE_ADD(CURDATE(), INTERVAL -1 DAY) <= `created_at`
```

Forcing a filter here would be counterproductive, since we might cause some records to be missed/skipped.

Perhaps we could have another event source just for these cases? Similar to the [Snowflake Query Results event source](https://github.com/PipedreamHQ/pipedream/tree/master/components/snowflake/sources/query-results)

Reply:
Updated to allow any custom query, and uses a column to dedupe.

Reply:
Very nice, I like the approach 🙂